src/solvers/refinement/string_constraint.h

```diff
@@ -0,0 +1,192 @@
+/*******************************************************************\
+
+Module: String constraints.
+        These are formulas talking about strings. We implemented two
+        forms of constraints: `string_constraintt` implements formulas
+        of the form $\forall univ_var \in [lb,ub[. premise => body$,
+	and not_contains_constraintt implements those of the form:
+        $\forall x in [lb,ub[. p(x) => \exists y in [lb,ub[.
+	s1[x+y] != s2[y]$.
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_H
+#define CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_H
+
+#include <langapi/language_ui.h>
+#include <solvers/refinement/bv_refinement.h>
+#include <solvers/refinement/refined_string_type.h>
+
+class string_constraintt: public exprt
+{
+public:
+  // String constraints are of the form
+  // forall univ_var in [lower_bound,upper_bound[. premise => body
+  // or premise => body
+
+  inline const exprt &premise() const
+  {
+    return op0();
+  }
+
+  inline const exprt &body() const
+  {
+    return op1();
+  }
+
+  inline const exprt &univ_var() const
+  {
+    return op2();
+  }
+
+  inline const exprt &upper_bound() const
+  {
+    return op3();
+  }
+
+  inline const exprt &lower_bound() const
+  {
+    return operands()[4];
+  }
+
+  // Trivial constraint
+  string_constraintt() : exprt(ID_string_constraint)
+  {
+    assert(false); // string constraints should not be initialized directly
+    copy_to_operands(true_exprt(), true_exprt());
+  }
+
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_inf,
+    const exprt &bound_sup,
+    const exprt &prem,
+    const exprt &body)
+    : exprt(ID_string_constraint)
+  {
+    copy_to_operands(prem, body);
+    copy_to_operands(univ, bound_sup, bound_inf);
+  };
+
+  // Default bound inferior is 0
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_sup,
+    const exprt &prem,
+    const exprt &body)
+    : string_constraintt(univ, refined_string_typet::index_zero(),
+                         bound_sup, prem, body)
+  {};
+
+  // Default premise is true
+  string_constraintt
+  (const symbol_exprt &univ, const exprt &bound_sup, const exprt &body)
+    : string_constraintt
+      (univ, refined_string_typet::index_zero(), bound_sup, true_exprt(), body)
+  {};
+
+  bool is_simple() const { return (operands().size()==2); }
+  bool is_univ_quant() const { return (operands().size()==5); }
+  bool is_not_contains() const { return false; }
+
+  inline symbol_exprt get_univ_var() const
+  { return to_symbol_expr(univ_var()); }
```

Comment:
No good reason. I have changed the code to only keep `univ_var` which now returns a `simbol_exprt`: https://github.com/diffblue/cbmc/pull/374/commits/22dae4d4cb39c7b5d4c7c28c0f7585e3224ea597